sklearn/preprocessing/tests/test_target_encoder.py

```diff
@@ -733,3 +735,185 @@ def test_pandas_copy_on_write():
+    """Force the reference (vectorized) transform path by disabling small-batch"""
+    old_thresh = enc._small_batch_threshold
+    enc._small_batch_threshold = -1
+    try:
+        return enc.transform(X)
+    finally:
+        enc._small_batch_threshold = old_thresh
+
+
+def test_norm_key_real_values_cover_nan_nat_and_strings():
+    """
+    Check that `_norm_key` maps NA-like values (None, float NaN, datetime/timedelta
+    NaT) to stable sentinel objects while leaving ordinary strings unchanged, and
+    that the sentinels for each NA family are distinct from one another.
+    """
+    k_none = _norm_key(None)
+
+    k_nan1 = _norm_key(float("nan"))
+    k_nan2 = _norm_key(np.float64(np.nan))
+    assert k_nan1 is k_nan2
+
+    k_str = _norm_key("x")
+    assert k_str == "x"
+
+    k_nat_dt = _norm_key(np.datetime64("NaT"))
+    k_nat_td = _norm_key(np.timedelta64("NaT"))
+    assert k_nat_dt is k_nat_td
+
+    assert k_none is not k_nan1 and k_none is not k_nat_dt
+    assert k_nan1 is not k_nat_dt and k_nan1 != "x"
+    assert k_nat_dt != "x"
+
+
+def test_small_batch_fastpath_matches_vectorized_with_nans_and_nats():
+    """
+    For tiny inputs, small-batch dict-lookup path should match the reference
+    vectorized path exactly (within tight numerical tolerance).
+    """
+    te_fast, te_vec = _fit_pair_numpy()
+
+    X = np.empty((5, 2), dtype=object)
+    X[0] = ["u", "x"]
+    X[1] = [float("nan"), "x"]
+    X[2] = [None, "x"]
+    X[3] = ["u", np.datetime64("NaT")]
+    X[4] = ["new_unseen", "x"]
+
+    npt.assert_allclose(te_fast.transform(X), te_vec.transform(X), rtol=0, atol=1e-9)
+
+
+def test_large_batch_vectorized_consistency():
+    """
+    For large inputs, both encoders choose the vectorized path; results must
+    match exactly (within tight tolerance).
+    """
+    te_fast, te_vec = _fit_pair_numpy()
+
+    base = np.array(
+        [["u", "x"], ["v", "y"], ["u", "x"], ["w", "y"], [None, "x"], ["new", "y"]],
+        dtype=object,
+    )
+    X_big = np.tile(base, (300, 1))
+
+    Z_fast = te_fast.transform(X_big)  # chooses vectorized due to size
+    Z_vec = te_vec.transform(X_big)  # always vectorized
+    npt.assert_allclose(Z_fast, Z_vec, rtol=0, atol=1e-9)
+
+
+@pytest.mark.parametrize("n_small", [1, 2, 8, 32])
+@pytest.mark.parametrize("n_cats", [10_000, 50_000])
+def test_small_batch_binary_parity_and_missing(n_small, n_cats):
+    rng = np.random.default_rng(0)
+    X_fit = rng.integers(0, n_cats, size=(150_000, 1)).astype(object)
+    # sprinkle missing during fit to learn its encoding too
+    X_fit[:100, 0] = None
+    y_fit = rng.integers(0, 2, size=(150_000,))
+    enc = TargetEncoder(random_state=0).fit(X_fit, y_fit)
+
+    X_small = rng.integers(0, n_cats * 2, size=(n_small, 1)).astype(object)
+    # add missing/unseen explicitly
+    if n_small >= 2:
+        X_small[0, 0] = None
+        X_small[-1, 0] = np.nan
+
+    ref = _force_slow_transform(enc, X_small)
```

Comment:
Just enforce the condition directly here without using this additional function.